### ISSUE TITLE
understanding the reco loss

### ISSUE THREAD

**Issue by seyeeet** [lorenmt/reco#7]
I am kinda new to contrast learning and although I completely follow the math, I have a confusion when I am looking at the code (by code I mean the general code, not only your code :) ).

So in the paper Eq.1 is the reco loss, which is the pixel wise contrastive loss, but in the code [here](https://github.com/lorenmt/reco/blob/3e856a316cd082f2e84bd3972d5f6c5fe24254bd/module_list.py#L139) I cannot understand why it is computed different from the way that Eq1 is written, is it a common thing in practice? 

In more details, in the code, we have `all_feat`, which is `256x513x256`, so for each sample the first one is the positive and the rest are the negatives.
The `seg_logits` then will have `256x513` dimension. We then compute the similarity via `cosine_similarity`  (which I think is different from the contrastive loss in eq 1).

Then, in the next step based on the `F.cross_entropy` we want all of them to have label zeros, because they have to be the same as the positive one.
My confusion is that why we dont use Eq1 as it is written in the paper and why we use `F.cross_entropy`.
I feel this is not exactly the same Eq1.
Can you please help me understand the relation?

**Comment by seyeeet** [comment]
Thank you very much for the explanation,
can you please tell me what part of the code corresponds to the numerator of equation 1 and what part of the code correspond to the denominator of the equation 1?
